packages/gatsby/lib/query-runner/route-writer.js

```diff
@@ -0,0 +1,255 @@
+  let splitChildRoutes = ``
+
+  const genChildRoute = (page, noPath) => {
+    let pathStr = ``
+    if (!noPath) {
+      pathStr = `path:'${page.path}',`
+    }
+    return `
+      {
+        ${pathStr}
+        component: ${page.internalComponentName},
+      },
+    `
+  }
+
+  const genSplitChildRoute = (page, noPath = false) => {
+    const pathName = pathChunkName(page.path)
+    const layoutName = layoutComponentChunkName(
+      program.directory,
+      page.component
+    )
+    let pathStr = ``
+    if (!noPath) {
+      if (program.prefixLinks) {
+        pathStr = `path:'${_.get(config, `linkPrefix`, ``)}${page.path}',`
+      } else {
+        pathStr = `path:'${page.path}',`
+      }
+    }
+
+    return `
+      { ${pathStr}
+        getComponent (nextState, cb) {
+          require.ensure([], (require) => {
+            const Component = preferDefault(require('${page.component}'))
+            require.ensure([], (require) => {
+              const data = require('./json/${page.jsonName}')
+              cb(null, () => <Component {...nextState} {...data} />)
+            }, '${pathName}')
+          }, '${layoutName}')
+        }
+      },
+    `
+  }
+
+  // Group pages under their layout component (if any).
+  let defaultLayoutExists = true
+  if (glob.sync(`${program.directory}/src/layouts/default.*`).length === 0) {
+    defaultLayoutExists = false
+  }
+  const groupedPages = _.groupBy(pages, page => {
+    // If is a string we'll assume it's a working layout component.
+    if (_.isString(page.layout)) {
+      return page.layout
+      // If the user explicitely turns off layout, put page in undefined bucket.
+    } else if (page.layout === false) {
+      return `undefined`
+      // Otherwise assume the default layout should handle this page.
+    } else if (defaultLayoutExists) {
+      return `default`
+      // We got nothing left, undefined.
+    } else {
+      return `undefined`
+    }
+  })
+  let rootRoute = `{ childRoutes: [` // We close this out at the bottom.
+  let splitRootRoute = `{ childRoutes: [` // We close this out at the bottom.
+  _.forEach(groupedPages, (pages, layout) => {
+    let route = `{`
+    let splitRoute = `{`
+    if (layout === `undefined`) {
+      // If a layout isn't defined then don't provide an index route, etc.
+      route += `childRoutes: [`
+      splitRoute += `childRoutes: [`
+      pages.forEach(page => {
+        route += genChildRoute(page)
+        splitRoute += genSplitChildRoute(page)
+      })
+      route += `]},`
+      splitRoute += `]},`
+      rootRoute += route
+      splitRootRoute += splitRoute
+    } else {
+      let indexPage
+      indexPage = pages.find(
+        page => parseFilepath(page.component).name === `index`
+      )
+
+      // If there's not an index page, just pick the one with the shortest path.
+      // Probably a bad heuristic.
+      if (!indexPage) {
+        indexPage = _.first(_.sortBy(pages, page => page.path.length))
```

Comment:
Just a style suggestion:
```javascript
_(pages).sortBy(page => page.path.length).first()
```
Quite same, but you retrieve the natural reading way.